Document what Nix *is* (#6420)
NixOS/nix: doc/manual/src/architecture/architecture.md

# Architecture

This chapter describes how Nix works.
It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.

## Overview

> Nix consists of hierarchical [layers](https://en.m.wikipedia.org/wiki/Multitier_architecture#Layers).

Reviewer: Links to mobile version of Wikipedia (here and in the following) probably unintentional?

Reviewer 2: Mobile Wikipedia is intentional, as it is less visual noise and page data.

                   [ configuration language  ]        |
                                |                     V
+-------------------------------|---------------------------------+
| store                         |                                 |
|                               | evaluates to                    |
|                               |                                 |
|               referenced by   V       builds                    |
|   [ build input ] ----> [ build plan ] ----> [ build result ]   |
|                                                                 |
+-----------------------------------------------------------------+
```

At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.

Below that is the [Nix expression language](../expressions/expression-language.md), a [purely functional](https://en.m.wikipedia.org/wiki/Purely_functional_programming) configuration language.
It is used to compose expressions which ultimately evaluate to self-contained *build plans*, used to derive *build results* from referenced *build inputs*.

::: {.note}

> The Nix language itself does not have a notion of *packages* or *configurations*.

Reviewer 3: I've drafted a definition for *package* in https://github.com/NixOS/nix/issues/6507. It is a bit more minimal and more general than the de facto handling of "packages" by the Nix tools.
Nonetheless, it is not part of the language per se, as it is a convention or interface around expressions.

As far as we are concerned here, the inputs and results of a derivation are just data.
In practice this amounts to a set of files in a file system.
:::

The command line and Nix language are what users interact with most.

> Underlying everything is the [Nix store](./store/store.md), a mechanism to keep track of build plans, data, and references between them.

Reviewer 2: [As you say](https://github.com/NixOS/nix/pull/6420#discussion_r869366058), all of that is in the package domain, which is merely a conventional structure within the Nix language, and one of the key features of `nixpkgs`. This is the manual for Nix proper, and right here and at that level of abstraction we would not care about these details.

What you add is true, so maybe you have a suggestion where to put that more appropriately?

> Underlying everything is the [Nix store](./store/store.md), a mechanism to keep track of build plans, data, and references between them.

Reviewer 3: I'd recommend to document the architecture of the store and language in separate sections/chapters/texts, with occasional cross-links to help readers see both the bigger picture and the implementation details.

Specifically, it seems like a good idea to use `::: {.note}` when talking about how the subject matter is used in a broader context (the reverse dependencies, so to speak) and regular links when referring to things that are used to implement the subject matter.

The issue on this line can be resolved by rephrasing it to be less absolute.

```suggestion
Underlying these is the [Nix store](./store/store.md), a mechanism to keep track of build plans, data, and references between them.
```

... which doesn't suggest as strongly that everything within these components would use it.

As for where my extra info should go, it's probably ok to relegate that entirely to the language documentation.

> Underlying everything is the [Nix store](./store/store.md), a mechanism to keep track of build plans, data, and references between them.

Reviewer 2: Thanks for the suggestion, merged.

> Specifically, it seems like a good idea to use ::: {.note} when talking about how the subject matter is used in a broader context (the reverse dependencies, so to speak) and regular links when referring to things that are used to implement the subject matter.

Interesting idea, noted.

# Architecture

This chapter describes how Nix works.

> It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.

Reviewer 2: >  if you want to do anything more with Nix than have a neat roll-back-able package manager, understanding the basic concepts that this document provides is critical

I thought so, too. After discussing with experienced people a lot, and doing studies with actual users, I have to disagree. We instead need better learning material, and this is why [we will put work into nix.dev](https://discourse.nixos.org/t/2022-07-07-documentation-team-meeting-notes-4/20198). I'm still convinced we have to front-load learners with specific knowledge instead of handing them inscrutable examples and templates, as Nix is different from what they usually know. But it has to be the right things, and user-oriented knowledge is different from the conceptual background presented here.

> To me, this document is about using and understanding Nix and its tooling

There should be some very condensed variant of this in the nix.dev guides. Right now it's not a priority for me. Would you be interested in giving it a try? The depth presented here should also be available on the onboarding path, but clearly optional. I'm convinced it's not really needed to use Nix effectively, but eventually studies will have to answer that question.

> If you want to start hacking on Nix itself or take inspiration for your own tools, I think there's a lot more to know.

Yes, absolutely. This architecture chapter should be the highest level of abstraction to start with. Eventually we should have paths into the source code, with additional prose sections discussing more detailed concepts, and then linking to the code to help discovering the implementation.

## Overview

Nix consists of [hierarchical layers][layer-architecture].

```
+-----------------------------------------------------------------+

> | Nix                                                             |

Reviewer 3: ```suggestion
| packaging system                                                |
```

Tbh this feels a bit absurd, but where do we draw the line of what terminology is "too specific" for the introduction?

By the same reasoning, we should perhaps replace `evaluates to` by `computes`?

Reviewer 2: > Tbh this feels a bit absurd, but where do we draw the line of what terminology is "too specific" for the introduction?

Exactly, this is why it is that way and nothing else.

> By the same reasoning, we should perhaps replace evaluates to by computes?

We use "evaluate" because that is what pure languages do. That term is also understood by people not into programming language theory, as it is self-explanatory - evaluation produces a value.

|                  [ commmand line interface ]------,             |
|                               |                   |             |
|                           evaluates               |             |
|                               |                manages          |
|                               V                   |             |
|                  [ configuration language  ]      |             |
|                               |                   |             |
| +-----------------------------|-------------------V-----------+ |

> | | store                  evaluates to                         | |

Reviewer 3: ```suggestion
| | process and data layer      |                                | |
| |                        evaluates to                         | |
```

Reviewer 2: Maybe we should rather rename the "persistence layer" in the table to "store"?

| |                             |                               | |
| |             referenced by   V       builds                  | |

> | |  [ build input ] ---> [ build plan ] ---> [ build result ]  | |

Reviewer 3: ```suggestion
| |  [ build input ] ---> [ build plan ] ---> [ outputs ]       | |
```

Isn't "output" the technical term for this, or should "result" encompass more than just the output? If so, what?
Example: `nix-store -q --outputs`

We should have a small glossary to which we then limit our use of terminology.

I wish markdown has `set -u` :wink:  (`nounset`)

Reviewer 2: > Isn't "output" the technical term for this

It's the Nix term for this, as is *derivation* for what we call *build task* here. We deliberately use generic terms in the overview, as none of that is specific to Nix.

> We should have a small glossary to which we then limit our use of terminology.

We have that as the [Rosetta Stone](https://github.com/NixOS/nix/pull/6420/files#diff-485e9f03f514869c78ebfa54354bd03b1fd0ba9ceb817578f31b573be303b5f4R181). The idea is to introduce Nix specific terminology in the subsequent sections, where we also describe how Nix implements the concepts in principle. We could move the table to the overview to accommodate newcomers and experienced Nix users, see https://github.com/NixOS/nix/pull/6420#discussion_r919878416.

Reviewer 3: > as none of that is specific to Nix.

Then `Nix` should be `package manager` and `store` should be `data and process layer`.

> Rosetta Stone

This doesn't scale to encompass terminology that is specific to Nix, so we should still have a glossary.

Would it make sense to have a translation of the entire diagram? Seems like a good way to learn the new terminology.

Reviewer 2: > Then `Nix` should be `package manager` and `store` should be `data and process layer`.

Sounds reasonable, please add a suggestion, considering the following:
- `data and process layer` reads a bit clunky, although it's the correct term
  - should we call it "store" generically instead, in the table?
- how do we still clarify this diagram refers to how Nix does things

> This doesn't scale to encompass terminology that is specific to Nix, so we should still have a glossary.

I hope this would be covered by additional sections, see [the colored outline comment above](https://github.com/NixOS/nix/pull/6420#issuecomment-1161690195). Is there anything that would still be missing?
Not against a glossary, it just needs enough substance. Maybe abbreviations such as drv or FSO, which we currently don't use and should not use in the official writing, but people may find helpful to decipher. On the other hand those may be NixOS Wiki material.

> Would it make sense to have a translation of the entire diagram? Seems like a good way to learn the new terminology.

Would take too much head space for my taste. Or maybe just link to the respective pages behind the table, not mentioning it in the table of contents? Sound like detail work though, not relevant for this PR.

Reviewer 3: Humans are really good at memorizing things in places, so I would guess that familiarizing oneself with the general diagram and then learning the technical terms in the same "place" would be highly effective. (I'm not a cognitive psychologist though)

You'd have to put them close together for this correspondence to be noticed; not one at the start and one at the end.

> Is there anything that would still be missing?

 - Terminology that is more specific, like "store path", "store hash", "store directory".
 - Terminology that builds on top of the core architecture, such as "input", which has a very specific meaning in Nixpkgs

One could argue that the latter is out of scope for this document, but in practice, all terminology throughout the ecosystem shares the same namespace. We should recognize that and have a single place to look all of them up, and disambiguate them where necessary. (e.g. "input", which is already conflicting with a term used in this chapter)

 > how do we still clarify this diagram refers to how Nix does things?
 
 The context would imply this, but it'd be helpful to have a caption such as "An overview of the Nix architecture using generic terminology", which then contrasts well with "An overview of the Nix architecture using its specific terms".

Reviewer 2: > Terminology that is more specific, like "store path", "store hash", "store directory".

We have that in the table of contents, and it is entirely Nix-specific.

> One could argue that the latter is out of scope for this document, but in practice, all terminology throughout the ecosystem shares the same namespace.

Agreed on potential confusion. Suggestions to do anything about that except always referring to "build input" in generic terms, because that is different from just "input"?

> You'd have to put them close together for this correspondence to be noticed; not one at the start and one at the end.

I agree with the learning conditions premise, but cannot picture a specific change. Feel free to make a pull request [on the fork](https://github.com/nix-community/) against this branch.

| |                                                             | |
| +-------------------------------------------------------------+ |
+-----------------------------------------------------------------+
```

At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.

Below that is the [Nix expression language](../expressions/expression-language.md), a [purely functional][purely-functional-programming] configuration language.

> It is used to compose expressions which ultimately evaluate to self-contained *build plans*, used to derive *build results* from referenced *build inputs*.

Reviewer 2: Agreed with the problem. I want to keep this essential section as concise as possible, though. Therefore definitely will not add variants in parentheses.

@tpwrules @Ericson2314 What about moving the Rosetta Stone to the end of the overview section?

The command line and Nix language are what users interact with most.

::: {.note}

> The Nix language itself does not have a notion of *packages* or *configurations*.
> As far as we are concerned here, the inputs and results of a derivation are just data.

Reviewer 2: I understand what you mean, and the qualification is sensible. I would require a different phrasing though, as I find "many uses of Nix" too vague. The problem is that anything I could come up with (e.g. "Although used for packaging and configuring software, ...") would make the sentence unwieldy without adding crucial information.

I would keep the correction w.r.t. terminology and leave it at that.
```suggestion
The Nix language itself does not have a notion of *packages* or *configurations*.
As far as we are concerned here, the inputs and results of a build plan are just data.
```

# Architecture

*(This chapter is unstable and a work in progress. Incoming links may rot.)*

> This chapter describes how Nix works.

Reviewer 4: This may sound like splitting hairs, but instead of the word "Nix", the manuals should say explicitly what the term refers to, even if it is to the detriment of brevity.<sup><b>1</b></sup> Depending on the context, I saw it used to mean:

+ Nix core concepts / paradigm / model
+ Nix CLI tools
+ Nix expression language
+ the implementation of Nix (i.e., the `NixOS/nix`)
+ the Nix ecosystem as a whole (i.e., `NixOS/nix` implementation, Nixpkgs, NixOS, even NixOps)
+ Nixpkgs (rarely, but it still)
+ (there is probably more..)

Taking the very next sentence as an example:
> It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.

+ "_NIx behaves_": I presume this alludes to the Nix model and/or the implementation
+ "_to modify Nix_": To modify a Nix expression, the behavior of the Nix CLI tools, the store object in the Nix store - or something else?

I guess the "Architecture" section is about the model itself, and if so, it would be prudent to start with a short explanation what it is about. (It would be nice to have a specification.) Below are two ways I tried to define "Nix model" to myself:

1. _A purely functional deployment model_
  ... then explain what purity<sup><b>2</b></sup>, "functional" ([this](https://github.com/NixOS/nix/blob/006e8f5b450c9524fe774b35954d342ee1c50d50/doc/manual/src/architecture/store/abstract/abstract.md) from PR #6877), and [deployment model](https://stackoverflow.com/a/73229950/1498178) mean.
  (Took it directly from the thesis; the omission of the word "software" is deliberate.)

2. _A (new) paradigm of building anything that requires assembly<sup><b>3</b></sup>, and managing the build results. The official term for "build result" is **store object** because they are kept in the Nix store._

---

\[1]: The definition of "Nix" has been a subject of contention on many online forums, but personally, it even made stop trying out Nix in the beginning and to give up on more certain parts of it a couple of times.  I found the tutorials, posts, manuals, etc. confusing because sometimes there were sudden context switches, and not knowing what other aspects of NIx there are, I couldn't connect the dots (heh, still can't most of the time).

\[2]: [Thesis](https://edolstra.github.io/pubs/phd-thesis.pdf), page 21 (PDF page 29), 2nd paragraph from bottom.

\[3]: My first try was "_A (new) paradigm of building digital artifacts<sup><b>3</b></sup>  (where the input can be anything that require assembly), and managing the build results._, but the term "digital artifact" seems just as overloaded as "component" or "package":
+ Model-Based Systems Engineering (MBSE) wiki: [Digital Artifact](https://www.omgwiki.org/MBSE/doku.php?id=mbse:digital_artifact)
+ Wikipedia: [Digital artifact](https://en.wikipedia.org/wiki/Digital_artifact)
+ WikiEducator: [Digital artefact](https://wikieducator.org/Digital_artefact)
+ Quora: [What is meant by digital artifact?](https://www.quora.com/What-is-meant-by-digital-artifact)

|                  [ commmand line interface ]------,             |
|                               |                   |             |
|                           evaluates               |             |
|                               |                manages          |
|                               V                   |             |
|                  [ configuration language  ]      |             |
|                               |                   |             |
| +-----------------------------|-------------------V-----------+ |
| | store                  evaluates to                         | |
| |                             |                               | |
| |             referenced by   V       builds                  | |
| |  [ build input ] ---> [ build plan ] ---> [ build result ]  | |
| |                                                             | |
| +-------------------------------------------------------------+ |
+-----------------------------------------------------------------+
```

At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.

> Below that is the [Nix expression language](../expressions/expression-language.md), a [purely functional][purely-functional-programming] configuration language.

Reviewer 4: When I started thinking about the term "Nix", this is the hierarchy that helped me put things into perspective:

```text

       ┌────────── Nix MODEL ──────────┐
       │                               │
       │ (This would be a spec, but    │
       │  it is the implementation     │
       │  itself, right?)              │
       │                               │
       └───────────────┬───────────────┘
                       │
                       │
                       │
                       ▼
┌───────────── Nix IMPLEMENTATION ──────────┐
│                 ("the core")              │
│                                           │    e.g., modules/c++ files
│         Implementation of the model       │          implementing flakes
│ (C++ code defining the building blocks of │
│  Nix: Nix lexer & parser, behaviour of    │
│  the Nix store, store objects and actions │
│  etc.)                                    │
│                                           │
└──────────────────────┬────────────────────┘
                       │
                       │
                       │
                       ▼
┌─────────────── Nix FRONTEND ──────────────┐            
│                                           │
│   + CLI tools that use the "core" libs    │    e.g., `nix flakes` commands
│                                           │
│   + Nix lang                              │
│     (Well, it is interpreted by the CLI   │
│      tools, but this is what is used to   │
│      interact with the system. Could be   │
│      wrong.)                              │
│                                           │
└───────────────────────────────────────────┘

```
